feat: add "unlisted" front matter option for blog posts

olafurpg opened the pull request:
Previously, the blog sidebar listed the most recent blog posts sorted by
their publish date. This commit adds the ability to hide a specific blog
post from the sidebar and the feed of all blog posts by adding the
`unlisted: true` header option.

Fixes #1393.

<!--
Thank you for sending the PR! We appreciate you spending the time to work on these changes.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to Docusaurus here: https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md

Happy contributing!

-->

## Motivation


The motivation for this feature is to make it possible to change dates of published posts without breaking the old URL (which might be widely shared).


Concrete example, we published a blog post two weeks ago that is dated for 2020 (https://scalameta.org/metals/blog/2020/04/12/mercury.html). We have closed three PRs fixing the date (https://github.com/scalameta/metals/pull/683, https://github.com/scalameta/metals/pull/670, https://github.com/scalameta/metals/pull/656) but they have all been closed because changing the date of the post would turn the old URL into a 404.

With the ability to unlist posts, I could unlist the 2020 post, create a new posts dated in 2019 and update the unlisted 2020 post to forward to the new 2019 post.

I'm open for alternative solutions to fix that particular problem. I think unlisting is a flexible solution that could be useful in other situations as well, for example when writing a draft post.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

I would appreciate help for testing this change since I'm unfamiliar with jest (and JavaScript in general 😅 ). I created a blog post in `__test__/__fixtures__` using `unlisted: true` and ran `yarn test` and everything passed but I expected some snapshot tests to fail.

I tested the change manually on our website and it seems to work as expected:

![2019-04-26 11 04 57](https://user-images.githubusercontent.com/1408093/56797045-186de280-6814-11e9-880a-b185b78a8ac5.gif)

Observe that uncommenting `unlisted: true` makes the post disappear from both the sidebar and the list of all blog posts while the post is still published under the unchanged URL.

## Related PRs

I'm not aware of any.

olafurpg commented on packages/docusaurus-1.x/lib/core/BlogPageLayout.js:
```diff
@@ -10,6 +10,8 @@ const BlogPost = require('./BlogPost.js');
 const Container = require('./Container.js');
 const MetadataBlog = require('./MetadataBlog.js');
+
+const MetadataPublicBlog = MetadataBlog.filter(item => item.unlisted !== true);
```
Where would be a good place to put this shared logic between `BlogSidebar.js`, `BlogPageLayout.js` and `feed.js`?

olafurpg commented on packages/docusaurus-1.x/lib/core/BlogPageLayout.js:
```diff
@@ -40,27 +42,28 @@ class BlogPageLayout extends React.Component {
           />
           <Container className="mainContainer postContainer blogContainer">
             <div className="posts">
-              {MetadataBlog.slice(page * perPage, (page + 1) * perPage).map(
```
This diff is easier on the eye with "No whitespace" enabled.

gabro commented on packages/docusaurus-1.x/lib/core/BlogPageLayout.js:
```diff
@@ -10,6 +10,8 @@ const BlogPost = require('./BlogPost.js');
 const Container = require('./Container.js');
 const MetadataBlog = require('./MetadataBlog.js');
+
+const MetadataPublicBlog = MetadataBlog.filter(item => item.unlisted !== true);
```
```suggestion
const MetadataPublicBlog = MetadataBlog.filter(item => !item.unlisted);
```

olafurpg replied:
Done.

NishealJ commented on packages/docusaurus-1.x/lib/server/feed.js:
```diff
@@ -28,6 +28,9 @@ module.exports = function(type) {
   readMetadata.generateMetadataBlog();
   const MetadataBlog = require('../core/MetadataBlog.js');
+  const MetadataPublicBlog = MetadataBlog.filter(
+    item => item.unlisted !== true,
```
item => !item.unlisted,

olafurpg replied:
Done.

olafurpg commented:
Thank you for the review! I've renamed "unlisted" into "draft" and incorporated the `!item.draft` refactoring. Note that there is a difference between draft posts in Jekyll and how they are implemented in this PR. 

- Jekyll draft posts are ignored during a non-development build and they do not need a date in the filename
- draft posts in this PR still require a date in the filename and are published during non-development builds, they just don't appear in the blog sidebar or main blog feed

> Is there currently a way for you to view your unlisted/draft posts without toggling the flag?

Not in the current implementation. It would be nice to see draft posts during livereload development but I'm not sure how to implement that.

gabro commented:
About unlisted vs draft. The way I would expect "draft" to behave, is not to make the post public, whereas unlisted would be published but not listed in the menu.

To summarize

 status  | visible in menu | published |
---------|:---------------:|:---------:|
draft    |      no         |   no      |
unlisted |      no         |   yes     |
normal   |      yes        |   yes     |

For this reason, I think 'unlisted' was the appropriate name for the changes included in this PR.

olafurpg commented:
Anything I can do to help unblock this issue? All of the review comments from https://github.com/facebook/Docusaurus/pull/1396#pullrequestreview-231255400 have been addressed, although I believe it's worth considering using `unlisted` instead of `draft` due to the reasons mentioned by @gabro in https://github.com/facebook/Docusaurus/pull/1396#issuecomment-487561180

endiliey commented:
cc @yangshun @wgao19
